src/Native/Unix/System.Native/pal_interfaceaddresses.c

```diff
@@ -195,15 +196,48 @@ int32_t SystemNative_EnumerateGatewayAddressesForInterface(uint32_t interfaceInd
         int isGateway = flags & RTF_GATEWAY;
         int gatewayPresent = hdr->rtm_addrs & RTA_GATEWAY;
 
-        if (isGateway && gatewayPresent)
+        if (isGateway && gatewayPresent && ((int)interfaceIndex == -1 || interfaceIndex == hdr->rtm_index))
         {
             IpAddressInfo iai;
+            struct sockaddr_storage* sock = (struct sockaddr_storage*)(hdr + 1);
```

Comment:
The data at `hdr + 1` can be interpreted as either a `sockaddr_storage` or a `sockaddr_in`?

Reply:
The sockaddr_storage is placeholder to be big enough to hold any socket family type.  They all overlap on length and family.

``` c 
struct sockaddr_storage {
  __uint8_t     ss_len;		/* address length */
  sa_family_t   ss_family;	/* [XSI] address family */
  char          __ss_pad1[_SS_PAD1SIZE];
  __int64_t     __ss_align;	/* force structure storage alignment */
  char          __ss_pad2[_SS_PAD2SIZE];
};
```

I could leave the sockaddr_in and cast it to sockaddr_in6 later but that looked strange to me. So I decided to use the placeholder type and cast to specific one after checking family type.